### PR TITLE
[RPA-961] Rework of Cloud Excel action: Set cells

### PR DESCRIPTION
**Pull request by bhszewczyk**
## Description

- Changed action name from "Update range" to "Set cells"
- Input changed from range to start cell address (to match desktop action version)
- Output not mandatory
- Allows user to set values to a range of cells
- Adds custom error handling
setCells

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

---

I agree to the terms of the RunBotics Contributor License Agreement.

### CODE REVIEW

File: runbotics/runbotics-desktop/src/action/microsoft/excel/automation/cloud-excel.error-message.ts [hunk 1]
```diff
@@ -16,6 +16,15 @@ export class CloudExcelErrorMessage {
             ${this.getCheckInputTip()}
         `;
     }
+
+    static setCellsIncorrectInput(): string {
+        return `
+            Start cell addresses has to include a column letter and a row number (e.g. A1).
+            Values has to be in a format of array of rows, where each row is an array of values to set, (e.g. [["value A3", "value B3"],["value A4", "value B4"]]).
+            For single cell update use setCell action.
+            ${this.getCheckInputTip()}
```
Comment (star-in-the-jar) [inline runbotics/runbotics-desktop/src/action/microsoft/excel/automation/cloud-excel.error-message.ts]: would fall out (wypadałoby xd), add how to insert empty values/forget some cells (null/empty string/undefined)

File: runbotics/runbotics-desktop/src/action/microsoft/excel/automation/types.ts [hunk 2]
```diff
@@ -17,9 +17,9 @@ export type CloudExcelSetCellActionInput = {
     cell: string;
 };
 
-export type CloudExcelUpdateRangeActionInput = {
-    range: string;
-    values: string | (string | number | boolean)[][];
+export type CloudExcelSetCellsActionInput = {
+    startCell: string;
+    values: (string | number | boolean)[][];
```
Comment (star-in-the-jar) [inline runbotics/runbotics-desktop/src/action/microsoft/excel/automation/types.ts]: use ExcelCellValue type
move it to rb-common to make it usable across packages

Comment (star-in-the-jar) [reply runbotics/runbotics-desktop/src/action/microsoft/excel/automation/types.ts]: remember to adjust existing imports

File: runbotics/runbotics-desktop/src/action/microsoft/excel/excel.service.ts [hunk 3]
```diff
@@ -139,14 +140,29 @@ export class ExcelService {
...
+    setCells(session: ExcelSession, startCell: string, values: ExcelCellValue[][]) {
         if (!Array.isArray(values)) {
             values = JSON.parse(values) as ExcelCellValue[][];
         }
 
+        if (!values.length) {
+            throw new Error(CloudExcelErrorMessage.setCellsIncorrectInput());
+        }
+        
+        const startColumnLetter = startCell.match(/[A-Z]+/).toString();
```
Comment (star-in-the-jar) [inline runbotics/runbotics-desktop/src/action/microsoft/excel/excel.service.ts]: there's (or soon will be) 

export enum ActionRegex {
 EXCEL_COLUMN_NAME = '^[a-zA-Z]+$', // add/use this line
}

Comment (star-in-the-jar) [reply runbotics/runbotics-desktop/src/action/microsoft/excel/excel.service.ts]: imo we could even create ExcelRegex and keep there all excel patterns

File: runbotics/runbotics-desktop/src/action/microsoft/excel/excel.service.ts [hunk 4]
```diff
@@ -139,14 +140,29 @@ export class ExcelService {
...
+        const startColumnNumber = this.getColumnNumber(startColumnLetter);
+        const startRow = +startCell.match(/\d+/);
```
Comment (star-in-the-jar) [inline runbotics/runbotics-desktop/src/action/microsoft/excel/excel.service.ts]: same hare - add pattern to ExcelRegex enum
+
what is that '= +star...' ???

File: runbotics/runbotics-desktop/src/action/microsoft/excel/excel.service.ts [hunk 5]
```diff
@@ -139,14 +140,29 @@ export class ExcelService {
...
     /**
      * @see https://learn.microsoft.com/en-us/graph/api/range-update?view=graph-rest-1.0&tabs=javascript
+     * @description {string} url requires range address in a format of column letters, e.g. "A1:C3"
      */
-    setRange(session: ExcelSession, address: string, values: string | ExcelCellValue[][]) {
-        const url = `/worksheets/${session.worksheetName}/range(address='${address}')`;
-
+    setCells(session: ExcelSession, startCell: string, values: ExcelCellValue[][]) {
         if (!Array.isArray(values)) {
             values = JSON.parse(values) as ExcelCellValue[][];
```
Comment (star-in-the-jar) [inline runbotics/runbotics-desktop/src/action/microsoft/excel/excel.service.ts]: if input won't be array the JSON will throw an error. consider wrapping this in try/catch and throwing   Error(CloudExcelErrorMessage.setCellsIncorrectInput());

File: runbotics/runbotics-desktop/src/action/microsoft/excel/excel.service.ts [hunk 6]
```diff
@@ -139,14 +140,29 @@ export class ExcelService {
...
+        const startColumnNumber = this.getColumnNumber(startColumnLetter);
+        const startRow = +startCell.match(/\d+/);
+        
+        const endColumnNumber = startColumnNumber + values[0].length - 1;
+        const endColumnLetter = this.getColumnLetter(endColumnNumber);
+        const endRow = startRow + values.length - 1;
+
+        const address = `${startColumnLetter}${startRow}:${endColumnLetter}${endRow}`;
+        
+        const url = `/worksheets/${session.worksheetName}/range(address='${address}')`;
```
Comment (star-in-the-jar) [inline runbotics/runbotics-desktop/src/action/microsoft/excel/excel.service.ts:167]: tidy 👌

File: runbotics/runbotics-desktop/src/action/microsoft/excel/excel.service.ts [hunk 7]
```diff
@@ -219,6 +235,16 @@ export class ExcelService {
         return (column.length - 1) * 26 + (column.charCodeAt(column.length - 1) - 64);
     }
 
+    /**
+     * @description converts column number to a letter format
+     * @param {number} columnNumber column number to convert
+     * @returns {string} column in a letter format
+     */
```
Comment (star-in-the-jar) [inline runbotics/runbotics-desktop/src/action/microsoft/excel/excel.service.ts:259]: imo add: @example getColumnLetter(29) // 'AC'

File: runbotics/runbotics-desktop/src/action/microsoft/excel/excel.service.ts (hunk 4 above)
Comment (preludium) [inline runbotics/runbotics-desktop/src/action/microsoft/excel/excel.service.ts]: dont use `+` sign for number parsing. Use `Number()` or `parseInt()`

File: runbotics/runbotics-desktop/src/action/microsoft/excel/excel.service.ts [hunk 8]
```diff
@@ -219,6 +235,16 @@ export class ExcelService {
...
+    private getColumnLetter(columnNumber: number): string {
+        const letter = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ'[(columnNumber - 1) % 26];
+        return columnNumber >= 26 ? this.getColumnLetter(Math.floor(columnNumber / 26) - 1) + letter : letter;
```
Comment (star-in-the-jar) [inline runbotics/runbotics-desktop/src/action/microsoft/excel/excel.service.ts]: shouldn't be there columnNumber > 26? Now i'm counting for getColumnLetter(26) // expected: 'Z' got: 'ZA'

Comment (star-in-the-jar) [inline runbotics/runbotics-desktop/src/action/microsoft/excel/excel.service.ts]: this is cool!

File: runbotics/runbotics-orchestrator-ui/src/src-app/translations/en/process/process.json [hunk 9]
```diff
@@ -307,9 +307,11 @@
...
     "Process.Details.Modeler.Actions.SharePointExcel.SetCell.Content": "Content",
     "Process.Details.Modeler.Actions.SharePointExcel.SetCell.Cell": "Specific cell",
-    "Process.Details.Modeler.Actions.CloudExcel.UpdateRange.Label": "Update range of cells",
-    "Process.Details.Modeler.Actions.SharePointExcel.UpdateRange.Cell": "Specific cell",
-    "Process.Details.Modeler.Actions.SharePointExcel.UpdateRange.Values": "Values [variable]",
+    "Process.Details.Modeler.Actions.CloudExcel.SetCells.Label": "Set cells",
+    "Process.Details.Modeler.Actions.SharePointExcel.SetCells.StartCell": "Start cell",
+    "Process.Details.Modeler.Actions.SharePointExcel.SetCells.StartCell.Info": "Starting cell address (e.g. A3)",
+    "Process.Details.Modeler.Actions.SharePointExcel.SetCells.Values": "Values",
+    "Process.Details.Modeler.Actions.SharePointExcel.SetCells.Values.Info": "Array of rows, where each row is an array of values to set, e.g. [[\"value A3\", \"value B3\"],[\"value A4\", \"value B4\"]]. If value is \"\" the cell will be cleared. If value is null the cell will be skipped",
```
Comment (star-in-the-jar) [inline runbotics/runbotics-orchestrator-ui/src/src-app/translations/en/process/process.json]: why '\' ???

Comment (star-in-the-jar) [reply runbotics/runbotics-orchestrator-ui/src/src-app/translations/en/process/process.json]: imo use singlequotes for increased readiblity

Comment (star-in-the-jar) [reply runbotics/runbotics-orchestrator-ui/src/src-app/translations/en/process/process.json]: good there's note about clearing/skipping cell

Comment (star-in-the-jar) [reply runbotics/runbotics-orchestrator-ui/src/src-app/translations/en/process/process.json]: add space in the array after comma

File: runbotics/runbotics-orchestrator-ui/src/src-app/translations/pl/process/process.json [hunk 10]
```diff
@@ -311,9 +311,11 @@
     "Process.Details.Modeler.Actions.CloudExcel.SetCell.Label": "Dodaj zawartość do komórki",
     "Process.Details.Modeler.Actions.SharePointExcel.SetCell.Content": "Zawartość komórki",
     "Process.Details.Modeler.Actions.SharePointExcel.SetCell.Cell": "Adres komórki",
-    "Process.Details.Modeler.Actions.CloudExcel.UpdateRange.Label": "Aktualizuj zakres komórek",
-    "Process.Details.Modeler.Actions.SharePointExcel.UpdateRange.Cell": "Zakres komórek",
-    "Process.Details.Modeler.Actions.SharePointExcel.UpdateRange.Values": "Wartości [zmienna]",
+    "Process.Details.Modeler.Actions.CloudExcel.SetCells.Label": "Ustaw komórki",
+    "Process.Details.Modeler.Actions.SharePointExcel.SetCells.StartCell": "Komórka początkowa",
+    "Process.Details.Modeler.Actions.SharePointExcel.SetCells.StartCell.Info": "Adres komórki początkowej (np. A3)",
```
Comment (star-in-the-jar) [inline runbotics/runbotics-orchestrator-ui/src/src-app/translations/pl/process/process.json]: in desktop we're passing examples after a comma:
"Adres komórki początkowej, np. A3"

**Review by star-in-the-jar — CHANGES_REQUESTED** [review]

File: runbotics/runbotics-desktop/src/action/microsoft/excel/automation/cloud-excel.error-message.ts [hunk 11]
```diff
@@ -16,6 +16,15 @@ export class CloudExcelErrorMessage {
             ${this.getCheckInputTip()}
         `;
     }
+
+    static setCellsIncorrectInput(): string {
+        return `
+            Start cell addresses has to include a column letter and a row number (e.g. A1).
```
Comment (preludium) [inline runbotics/runbotics-desktop/src/action/microsoft/excel/automation/cloud-excel.error-message.ts]: too long.
`StartCell field expects cell address e.g. A3.
Values field expects to specify nested lists. Each list represents a row and the elements inside the values in the corresponding column, e.g. [["A3 value", "B3 value"],["A4 value", "B4 value"]])`

File: runbotics/runbotics-desktop/src/action/microsoft/excel/automation/types.ts (hunk 2 above)
Comment (preludium) [inline runbotics/runbotics-desktop/src/action/microsoft/excel/automation/types.ts]: we already have a type for this union

File: runbotics/runbotics-desktop/src/action/microsoft/excel/excel.service.ts (hunk 3 above)
Comment (preludium) [inline runbotics/runbotics-desktop/src/action/microsoft/excel/excel.service.ts]: we should not expect uppercase letters to be provided, just make it uppercase here or match like this `/A-Za-z]+/`

File: runbotics/runbotics-desktop/src/action/microsoft/excel/excel.service.ts [hunk 12]
```diff
@@ -219,6 +235,16 @@ export class ExcelService {
         return (column.length - 1) * 26 + (column.charCodeAt(column.length - 1) - 64);
     }
 
+    /**
+     * @description converts column number to a letter format
+     * @param {number} columnNumber column number to convert
+     * @returns {string} column in a letter format
+     */
+    private getColumnLetter(columnNumber: number): string {
+        const letter = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ'[(columnNumber - 1) % 26];
```
Comment (preludium) [inline runbotics/runbotics-desktop/src/action/microsoft/excel/excel.service.ts]: move it to const outside of class `'ABCDEFGHIJKLMNOPQRSTUVWXYZ'`

**Review by preludium — CHANGES_REQUESTED** [review]

File: runbotics/runbotics-desktop/src/action/microsoft/excel/excel.service.ts (hunk 8 above)
Comment (bhszewczyk) [reply runbotics/runbotics-desktop/src/action/microsoft/excel/excel.service.ts]: thanks! :)

File: runbotics/runbotics-common/src/actions/actions.ts [hunk 13]
```diff
@@ -208,6 +210,8 @@ export enum DesktopAction {
 	READ_TEXT_FROM_IMAGE = 'desktop.readTextFromImage',
 }
 
+export type ExcelCellValue = string | number | boolean;
```
Comment (preludium) [inline runbotics/runbotics-common/src/actions/actions.ts]: why common? do we use it outside of bot package?

**Review by preludium — APPROVED** [review]

File: runbotics/runbotics-common/src/actions/actions.ts (hunk 13 above)
Comment (star-in-the-jar) [reply runbotics/runbotics-common/src/actions/actions.ts]: that's been my suggestion. We do use it in runbotics-actions

Comment (star-in-the-jar) [reply runbotics/runbotics-common/src/actions/actions.ts]: restoring then

File: runbotics/runbotics-desktop/src/action/microsoft/excel/automation/cloud-excel.error-message.ts [hunk 14]
```diff
@@ -16,6 +16,17 @@ export class CloudExcelErrorMessage {
...
         `;
     }
+
+    static setCellsIncorrectInput(): string {
+        return `
+            Start cell field expects cell address (e.g. A1).
+            Values field expects specified nested lists. Each list represents a row and the elements inside the values in the corresponding column, e.g. [["A3 value", "B3 value"], ["A4 value", "B4 value"]])
+            To skip a cell use null.
+            To clear a cell use "".
+            For single cell update use setCell action.
```
Comment (star-in-the-jar) [inline runbotics/runbotics-desktop/src/action/microsoft/excel/automation/cloud-excel.error-message.ts:35]: awesome ❤️

File: runbotics/runbotics-desktop/src/action/microsoft/excel/excel.service.ts [hunk 15]
```diff
@@ -22,6 +21,9 @@ import {
 } from './excel.types';
 import { OneDriveService } from '../one-drive';
 import { hasWorkbookSessionId, hasWorksheetName } from './excel.utils';
+import { CloudExcelErrorMessage } from './automation/cloud-excel.error-message';
+
+const alphabet = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ';
```
Comment (star-in-the-jar) [inline runbotics/runbotics-desktop/src/action/microsoft/excel/excel.service.ts]: it's a const co make it ALPHABET or optionaly GOOGLE

File: runbotics/runbotics-orchestrator-ui/src/src-app/translations/en/process/process.json [hunk 16]
```diff
@@ -307,9 +307,11 @@
     "Process.Details.Modeler.Actions.CloudExcel.SetCell.Label": "Add content to cell",
     "Process.Details.Modeler.Actions.SharePointExcel.SetCell.Content": "Content",
     "Process.Details.Modeler.Actions.SharePointExcel.SetCell.Cell": "Specific cell",
-    "Process.Details.Modeler.Actions.CloudExcel.UpdateRange.Label": "Update range of cells",
-    "Process.Details.Modeler.Actions.SharePointExcel.UpdateRange.Cell": "Specific cell",
-    "Process.Details.Modeler.Actions.SharePointExcel.UpdateRange.Values": "Values [variable]",
+    "Process.Details.Modeler.Actions.CloudExcel.SetCells.Label": "Set cells",
+    "Process.Details.Modeler.Actions.SharePointExcel.SetCells.StartCell": "Start cell",
+    "Process.Details.Modeler.Actions.SharePointExcel.SetCells.StartCell.Info": "Starting cell address (e.g. A3)",
```
Comment (star-in-the-jar) [inline runbotics/runbotics-orchestrator-ui/src/src-app/translations/en/process/process.json]: Starting cell address, e.g. A3.

**Review by star-in-the-jar — APPROVED** [review]

**Review by star-in-the-jar — APPROVED** [review]

**Pull request merged**